crates/binjs_io/src/escaped_wtf8.rs

```diff
@@ -0,0 +1,276 @@
+            buf.push(LONE_SORROGATE_UNIT_3_MIN + low);
+        }
+
+        // ...... \x7F XXXX ......
+        //                  ^
+        //                  |
+        //                begin
+        begin = end + 5;
+
+        let pos = input[begin..].iter().position(|&c| c == LONE_SORROGATE_ESCAPE_CHAR);
+        if pos.is_none() {
+            buf.extend_from_slice(&input[begin..]);
+            break;
+        }
+
+        // ...... \x7F XXXX ...... \x7F XXXX ......
+        //                  ^      ^
+        //                  |      |
+        //                begin   end
+        end = pos.unwrap();
+    };
+
+    Some(buf)
+}
+
+/// If the given `bytes` contains any escaped lone surropgate, convert it to
+/// \uXXXX format, just for debug print.
+/// This does unrecoverable conversion, the result shouldn't be used outside
+/// of debug print.
+pub fn for_print(s: &SharedString) -> SharedString {
+    let pos = s.as_str().find(char::from(LONE_SORROGATE_ESCAPE_CHAR));
+    if pos.is_none() {
+        return s.clone();
+    }
+
+    let input = s.as_str().as_bytes();
+
+    //   ...... \x7F XXXX ......
+    //   ^      ^
+    //   |      |
+    // begin   end
+    let mut begin = 0;
+    let mut end = pos.unwrap();
+
+    let mut buf: Vec<u8> = Vec::new();
+    loop {
+        let head = &input[begin..end];
+        buf.extend_from_slice(head);
+
+        let codepoint: u16 =
+            (decode_hex_char(input[end + 1]) << 12) |
+            (decode_hex_char(input[end + 2]) << 8) |
+            (decode_hex_char(input[end + 3]) << 4) |
+            decode_hex_char(input[end + 4]);
+        if codepoint == LONE_SORROGATE_ESCAPE_CHAR as u16 {
+            buf.push(LONE_SORROGATE_ESCAPE_CHAR);
+        } else {
+            assert!(codepoint >= LEAD_SORROGATE_MIN && codepoint <= TRAIL_SORROGATE_MAX,
+                    "escaped codepoint should be either {:04x} or lone surrogate ({:04x}...{:04x})",
+                    LONE_SORROGATE_ESCAPE_CHAR,
+                    LEAD_SORROGATE_MIN,
+                    TRAIL_SORROGATE_MAX);
+
+            buf.push(b'\\');
+            buf.push(b'u');
+            buf.push(input[end + 1]);
+            buf.push(input[end + 2]);
+            buf.push(input[end + 3]);
+            buf.push(input[end + 4]);
+        }
+
+        // ...... \x7F XXXX ......
+        //                  ^
+        //                  |
+        //                begin
+        begin = end + 5;
+
+        let pos = input[begin..].iter().position(|&c| c == LONE_SORROGATE_ESCAPE_CHAR);
+        if pos.is_none() {
+            buf.extend_from_slice(&input[begin..]);
+            break;
+        }
+
+        // ...... \x7F XXXX ...... \x7F XXXX ......
+        //                  ^      ^
+        //                  |      |
+        //                begin   end
+        end = pos.unwrap();
+    };
+
+    SharedString::from_string(String::from_utf8(buf).expect("Escaped string should be valid UTF-8"))
```

Comment:
I think this UTF-8 check could be avoided by making `buf` a `String` instead of `Vec<u8>`, using `push_str` instead of `extend_from_slice`, etc.

Reply:
it makes `input[...]` part complicated, and different from `unescape`, so I'd like to keep it as is.

Reply:
Sorry, what `input[...]` part? `str` can be sliced with range indexing syntax too. Anyway, as you prefer. These are just suggestions, as I said before feel free to ignore any :)